pkg/frontend/transport/roundtripper.go

```diff
@@ -43,22 +44,40 @@ func (a *grpcRoundTripperAdapter) RoundTrip(r *http.Request) (*http.Response, er
 		return nil, err
 	}
 
-	resp, err := a.roundTripper.RoundTripGRPC(r.Context(), req)
+	resp, body, err := a.roundTripper.RoundTripGRPC(r.Context(), req)
 	if err != nil {
 		var ok bool
 		if resp, ok = httpgrpc.HTTPResponseFromError(err); !ok {
 			return nil, err
 		}
 	}
 
+	var respBody io.ReadCloser
+	if body != nil {
+		respBody = body
+	} else {
+		respBody = &buffer{buff: resp.Body, ReadCloser: io.NopCloser(bytes.NewReader(resp.Body))}
+	}
+
 	httpResp := &http.Response{
-		StatusCode:    int(resp.Code),
-		Body:          &buffer{buff: resp.Body, ReadCloser: io.NopCloser(bytes.NewReader(resp.Body))},
-		Header:        http.Header{},
-		ContentLength: int64(len(resp.Body)),
+		StatusCode: int(resp.Code),
+		Body:       respBody,
+		Header:     http.Header{},
 	}
 	for _, h := range resp.Headers {
 		httpResp.Header[h.Key] = h.Values
 	}
+
+	contentLength := -1
+	if len(resp.Body) > 0 {
+		contentLength = len(resp.Body)
+	} else if l := httpResp.Header.Get("Content-Length"); l != "" {
+		cl, err := strconv.Atoi(l)
+		if err != nil {
+			contentLength = cl
+		}
```

Comment:
wouldn't this already be taken care of in the loop above? `	for _, h := range resp.Headers {...}`

Reply:
I agree that it should be sufficient to set the field. However, the previous logic was setting field and header, so I thought it would be least surprising to keep the same logic and set both.

Reply:
>I agree that it should be sufficient to set the field. However, the previous logic was setting field and header, so I thought it would be least surprising to keep the same logic and set both, wdyt?

makes sense 👍 let's keep it that way